Improve build file syntax error

tgolsson opened the pull request:
Inspired by this thread: https://pantsbuild.slack.com/archives/C046T6T9U/p1706779724173179

What it currently looks like: 
<img width="805" alt="image" src="https://github.com/pantsbuild/pants/assets/8234817/33a8ed1f-e019-48dc-b1c2-c92a5122a3d1">

With this change:
<img width="908" alt="image" src="https://github.com/pantsbuild/pants/assets/8234817/886de937-c1d9-4d41-b168-305f3b1cf5da">

In later Python versions we would also have end_lineno and end_offset, but those properties aren't accessible (or set?) in 3.9.

kaos reviewed (commented):
Seeing we rely on the details of the `SyntaxError` class, I'd love to have a simple test that verifies the resulting error message so we don't get an unchecked regression when upgrading our Python version in the future.

tgolsson commented:
> Seeing we rely on the details of the `SyntaxError` class, I'd love to have a simple test that verifies the resulting error message so we don't get an unchecked regression when upgrading our Python version in the future.

Added a test, and reworked the raise mechanic. I think using `raise ... from ...` would've led to confusion when seeing both errors. Now it will be a continuous callstack, we just replace the exception value when it leaves the standard library.

kaos reviewed (approved):
Great

huonw commented:
So good! Nice one
